src/app.rs

```diff
@@ -167,27 +177,94 @@ async fn schema() -> &'static str {
 ///
 /// * `client`: S3 client object
 /// * `request_data`: RequestData object for the request
-#[tracing::instrument(
-    level = "DEBUG",
-    skip(client, request_data, resource_manager, mem_permits)
-)]
-async fn download_object<'a>(
+/// * `resource_manager`: ResourceManager object
+async fn download_s3_object<'a>(
     client: &s3_client::S3Client,
     request_data: &models::RequestData,
     resource_manager: &'a ResourceManager,
-    mem_permits: &mut Option<SemaphorePermit<'a>>,
 ) -> Result<Bytes, ActiveStorageError> {
+
+    // If we're given a size in the request data then use this to
+    // get an initial guess at the required memory resources.
+    let memory = request_data.size.unwrap_or(0);
+    let mut mem_permits = resource_manager.memory(memory).await?;
+
     let range = s3_client::get_range(request_data.offset, request_data.size);
     let _conn_permits = resource_manager.s3_connection().await?;
+
     client
+    .download_object(
+        &request_data.bucket,
+        &request_data.object,
+        range,
+        resource_manager,
+        &mut mem_permits,
+    )
+    .await
+}
+
+/// Download and cache an object from S3
+///
+/// Requests a byte range if `offset` or `size` is specified in the request.
+///
+/// # Arguments
+///
+/// * `client`: S3 client object
+/// * `request_data`: RequestData object for the request
+/// * `resource_manager`: ResourceManager object
+/// * `chunk_cache`: ChunkCache object
+async fn download_and_cache_s3_object<'a>(
+    client: &s3_client::S3Client,
+    request_data: &models::RequestData,
+    resource_manager: &'a ResourceManager,
+    chunk_cache: &ChunkCache,
+) -> Result<Bytes, ActiveStorageError> {
+
+    let key = format!("{},{:?}", client, request_data);
+
+    match chunk_cache.get(&key).await {
+        Ok(value) => {
+            if let Some(bytes) = value {
+                return Ok(bytes);
+            }
+        },
+        Err(e) => {
+            return Err(e);
+        }
+    }
+
+    // If we're given a size in the request data then use this to
+    // get an initial guess at the required memory resources.
+    let memory = request_data.size.unwrap_or(0);
+    let mut mem_permits = resource_manager.memory(memory).await?;
+
+    let range = s3_client::get_range(request_data.offset, request_data.size);
+    let _conn_permits = resource_manager.s3_connection().await?;
+
+    let data = client
         .download_object(
             &request_data.bucket,
             &request_data.object,
             range,
             resource_manager,
-            mem_permits,
+            &mut mem_permits,
         )
-        .await
+        .await;
```

Comment:
Can we replace this with a call to the new `download_s3_object` function for DRY-ness?